src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,391 @@
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    HkdfKeyDerivation(String hmacAlgName, int hmacLen,
+                      AlgorithmParameterSpec algParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        super(algParameterSpec);
+        if (algParameterSpec != null) {
+            throw new InvalidAlgorithmParameterException(
+                "RFC 5869 has no parameters for its KDF algorithms");
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived, or specifies
+     *     a type of output that is not a key (e.g. raw data)
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+
+        if (alg == null || alg.isEmpty()) {
+            throw new InvalidParameterSpecException(
+                "the algorithm for the resultant SecretKey may not be null or"
+                + " empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(kdfParameterSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived key material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        SecretKey inputKeyMaterial;
+        SecretKey salt;
+        SecretKey pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (kdfParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) kdfParameterSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
```

Comment:
There is a bug here. `consolidateKeyMaterial()` returns the key if it's the only one in `ikms`, without checking if it's "bogus". Then, it is passed into `hkdfExtract` and very unfortunately `hmacObj.doFinal(null)` doesn't make any noise.

Same in ExtractThenExpand.